fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreBase.java

```diff
@@ -174,16 +174,48 @@ default FDBStoreTimer getTimer() {
+         * index must be built in the same transaction that opens the record store, which can lead to errors
+         * on large stores if the index cannot be built in the
+         * <a href="https://apple.github.io/foundationdb/known-limitations.html#long-running-transactions">five second
+         * FoundationDB transaction time limit</a>.
+         *
+         * <p>
+         * By default, this will return {@link IndexState#READABLE} for any indexes on new types (which
+         * can be used right away without doing any I/O) or if the number of records in the store is small (below
+         * {@link FDBRecordStore#MAX_RECORDS_FOR_REBUILD}). However, if the record store is large, this will return
+         * {@link IndexState#DISABLED}, which indicates that the index should not be maintained and that it cannot be
+         * used until the index is built by the {@link OnlineIndexer}.
+         * </p>
+         *
+         * <p>
+         * For adopters, two utility methods are provided that can be used to make implementing this method easier.
+         * The first is {@link FDBRecordStore#disabledIfTooManyRecordsForRebuild(long, boolean)}, which replicates
+         * the default behavior. The second is
+         * {@link FDBRecordStore#writeOnlyIfTooManyRecordsForRebuild(long, boolean)}, which is similar to the default
+         * except that it returns {@link IndexState#WRITE_ONLY} instead of {@link IndexState#DISABLED} and was the
+         * default prior to Record Layer 3.0. Note that all indexes must be made {@link IndexState#WRITE_ONLY} before
+         * they can be built, but the {@link OnlineIndexer} should generally handle that index state transition, and so
+         * most adopters should return {@link IndexState#DISABLED} on indexes that cannot be built in-line.
+         * </p>
+         *
          * @param index the index that has not been built for this store
          * @param recordCount the number of records already in the store
          * @param indexOnNewRecordTypes <code>true</code> if all record types for the index are new (the number of
          *                              records related to this index is 0), in which case the index is able to be
          *                              "rebuilt" instantly with no cost.
          * @return the desired state of the new index. If this is {@link IndexState#READABLE}, the index will be built right away
+         * @see FDBRecordStore#disabledIfTooManyRecordsForRebuild(long, boolean)
+         * @see FDBRecordStore#writeOnlyIfTooManyRecordsForRebuild(long, boolean)
          */
         default IndexState needRebuildIndex(Index index, long recordCount, boolean indexOnNewRecordTypes) {
-            return FDBRecordStore.writeOnlyIfTooManyRecordsForRebuild(recordCount, indexOnNewRecordTypes);
+            return FDBRecordStore.disabledIfTooManyRecordsForRebuild(recordCount, indexOnNewRecordTypes);
```

Comment:
Is the `index` arg ever used? (couldn't find a usage in `UserVersionChecker`'s implementations too). 
If not, should it be removed?

Reply:
It's not used in here or in the Record Layer codebase itself, but this is an interface that adopters might themselves implement, and the index argument is there so that they could make different decisions about different indexes. (For example, you could imagine someone wanting to have "experimental" indexes return `DISABLED` all the time, even on new stores or new types. In that case, you could implement a `UserVersionChecker` that somehow identified the index as experimental (perhaps by the `IndexOptions` on that index) and then always returned `IndexState.DISABLED` regardless of the values of the other two parameters on such indexes.)